public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts

```diff
@@ -2,72 +2,105 @@ import _ from 'lodash';
-      if (aiPromise) {
-        promises.push(aiPromise);
+    const byType: Record<AzureQueryType, DataQueryRequest<AzureMonitorQuery>> = ({} as unknown) as Record<
+      AzureQueryType,
+      DataQueryRequest<AzureMonitorQuery>
+    >;
+
+    for (const target of options.targets) {
+      // Migrate old query structure
+      if (target.queryType === AzureQueryType.ApplicationInsights) {
+        if ((target.appInsights as any).rawQuery) {
+          target.queryType = AzureQueryType.InsightsAnalytics;
+          target.insightsAnalytics = (target.appInsights as unknown) as InsightsAnalyticsQuery;
+          delete target.appInsights;
+        }
       }
-    }
-
-    if (azureLogAnalyticsOptions.targets.length > 0) {
-      const obs = this.azureLogAnalyticsDatasource.query(azureLogAnalyticsOptions);
-      if (!promises.length) {
-        return obs; // return the observable directly
+      if (!target.queryType) {
+        target.queryType = AzureQueryType.AzureMonitor;
       }
-      // NOTE: this only includes the data!
-      // When all three query types are ready to be observale, they should all use observable
-      promises.push(obs.toPromise().then(r => r.data));
-    }
 
-    if (azureMonitorOptions.targets.length > 0) {
-      const obs = this.azureMonitorDatasource.query(azureMonitorOptions);
-      if (!promises.length) {
-        return obs; // return the observable directly
+      // Check that we have options
+      const opts = (target as any)[this.optionsKey[target.queryType]];
+
+      // Skip hidden queries or ones without properties
+      if (target.hide || !opts) {
+        continue;
       }
-      // NOTE: this only includes the data!
-      // When all three query types are ready to be observale, they should all use observable
-      promises.push(obs.toPromise().then(r => r.data));
-    }
 
-    if (promises.length === 0) {
-      return Promise.resolve({ data: [] });
+      // Initalize the list of queries
+      let q = byType[target.queryType];
+      if (!q) {
+        q = _.cloneDeep(options);
+        q.targets = [];
+        byType[target.queryType] = q;
+      }
+      q.targets.push(target);
     }
 
-    return Promise.all(promises).then(results => {
-      return { data: _.flatten(results) };
+    // Distinct types are managed by distinct requests
+    const queries = Object.keys(byType).map((type: AzureQueryType) => {
+      const req = byType[type];
+      return this.pseudoDatasource[type].query(req);
     });
+    // Single query can skip merge
+    if (queries.length === 1) {
+      return queries[0];
+    }
+    if (queries.length > 1) {
+      // Not accurate, but simple and works
+      // should likely be more like the mixed data source
+      const promises = queries.map(obs => obs.toPromise());
+      return Promise.all(promises).then(results => {
+        return { data: _.flatten(results) };
+      });
+    }
+    return of({ state: LoadingState.Done });
```

Comment:
For future PRs, do we need to return both Promises and Observable here? I think it makes the code more complicated. Maybe we should always return Observable?